{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "59da011f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/eddiekong/ml4fg/isoform-gen/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from pyfaidx import Fasta\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import pyarrow.dataset as ds\n",
    "\n",
    "from dataset import GeneIterableFixedB, split_filters_by_chroms\n",
    "import torch\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "278959d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "usecols = [0, 2, 3, 4, 6, 8]\n",
    "colnames = [\"chrom\", \"feature\", \"start\", \"end\", \"strand\", \"attributes\"]\n",
    "\n",
    "gtf = pd.read_csv(\n",
    "    \"data/raw_data/ENCFF129ZDE.gtf.gz\",\n",
    "    sep=\"\\t\",\n",
    "    comment=\"#\",\n",
    "    header=None,\n",
    "    names=colnames,\n",
    "    usecols=usecols,\n",
    "    compression=\"gzip\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e44467ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29321</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_name \"WASH7P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>15005</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18379</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     feature  start    end strand  \\\n",
       "12  chr1        gene  14404  29321      -   \n",
       "13  chr1  transcript  14404  29570      -   \n",
       "14  chr1        exon  29534  29570      -   \n",
       "15  chr1        exon  24738  24891      -   \n",
       "16  chr1        exon  18268  18366      -   \n",
       "17  chr1        exon  17915  18061      -   \n",
       "18  chr1        exon  17606  17742      -   \n",
       "19  chr1        exon  17233  17368      -   \n",
       "20  chr1        exon  16858  17055      -   \n",
       "21  chr1        exon  16607  16765      -   \n",
       "22  chr1        exon  15796  15947      -   \n",
       "23  chr1        exon  15005  15038      -   \n",
       "24  chr1        exon  14404  14501      -   \n",
       "25  chr1  transcript  14404  14501      -   \n",
       "26  chr1        exon  14404  14501      -   \n",
       "27  chr1  transcript  14404  24891      -   \n",
       "28  chr1        exon  24738  24891      -   \n",
       "29  chr1        exon  18268  18379      -   \n",
       "\n",
       "                                           attributes  \n",
       "12  gene_id \"ENSG00000227232.5\"; gene_name \"WASH7P...  \n",
       "13  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "14  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "15  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "16  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "17  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "18  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "19  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "20  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "21  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "22  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "23  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "24  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "25  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "26  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "27  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "28  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "29  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# row = gtf.iloc[1]\n",
    "# import pprint\n",
    "# pprint.pprint(row.to_dict(), width=120)\n",
    "\n",
    "gtf.iloc[12:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "193d8b24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129911\n"
     ]
    }
   ],
   "source": [
    "# read the Parquet file\n",
    "df = pd.read_parquet(\"data/table/ENCFF180NXA.parquet\")\n",
    "\n",
    "# number of rows\n",
    "print(len(df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cb37824f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrom': 'chr16',\n",
      " 'end': 57591681,\n",
      " 'gene_id': 'ENSG00000159618.15',\n",
      " 'start': 57542421,\n",
      " 'strand': '+',\n",
      " 'transcripts': array([{'transcript_id': 'ENST00000615867.4', 'tx_start': 57542687, 'tx_end': 57565313, 'exon_starts': array([57542687, 57562056, 57562384, 57563091, 57563848, 57565034]), 'exon_ends': array([57542701, 57562157, 57562459, 57563247, 57563979, 57565313]), 'rel_abundance': 1.0}],\n",
      "      dtype=object)}\n",
      "Transcripts: 351270\n"
     ]
    }
   ],
   "source": [
    "row = df.iloc[18197]\n",
    "pprint.pprint(row.to_dict(), width=120)\n",
    "\n",
    "num_transcripts = df[\"transcripts\"].apply(len).sum()\n",
    "\n",
    "print(\"Transcripts:\", num_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f4efcb4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matching genes: 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id, chrom, start, end]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target = \"ENCLB376UFGT000206877\"\n",
    "\n",
    "matches = df[df[\"transcripts\"].apply(\n",
    "    lambda txs: any(t[\"transcript_id\"] == target for t in txs)\n",
    ")]\n",
    "\n",
    "print(\"Number of matching genes:\", len(matches))\n",
    "matches[[\"gene_id\", \"chrom\", \"start\", \"end\"]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isoform-gen",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
