{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "59da011f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/eddiekong/ml4fg/isoform-gen/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from pyfaidx import Fasta\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import pyarrow.dataset as ds\n",
    "\n",
    "from dataset import GeneIterableFixedB, split_filters_by_chroms\n",
    "import torch\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "278959d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "usecols = [0, 2, 3, 4, 6, 8]\n",
    "colnames = [\"chrom\", \"feature\", \"start\", \"end\", \"strand\", \"attributes\"]\n",
    "\n",
    "gtf = pd.read_csv(\n",
    "    \"data/raw_data/ENCFF129ZDE.gtf.gz\",\n",
    "    sep=\"\\t\",\n",
    "    comment=\"#\",\n",
    "    header=None,\n",
    "    names=colnames,\n",
    "    usecols=usecols,\n",
    "    compression=\"gzip\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e44467ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29321</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_name \"WASH7P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>29534</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18366</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17915</td>\n",
       "      <td>18061</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17606</td>\n",
       "      <td>17742</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>17233</td>\n",
       "      <td>17368</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>16858</td>\n",
       "      <td>17055</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>16607</td>\n",
       "      <td>16765</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>15796</td>\n",
       "      <td>15947</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>15005</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>14404</td>\n",
       "      <td>14501</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>24738</td>\n",
       "      <td>24891</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>exon</td>\n",
       "      <td>18268</td>\n",
       "      <td>18379</td>\n",
       "      <td>-</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     feature  start    end strand  \\\n",
       "12  chr1        gene  14404  29321      -   \n",
       "13  chr1  transcript  14404  29570      -   \n",
       "14  chr1        exon  29534  29570      -   \n",
       "15  chr1        exon  24738  24891      -   \n",
       "16  chr1        exon  18268  18366      -   \n",
       "17  chr1        exon  17915  18061      -   \n",
       "18  chr1        exon  17606  17742      -   \n",
       "19  chr1        exon  17233  17368      -   \n",
       "20  chr1        exon  16858  17055      -   \n",
       "21  chr1        exon  16607  16765      -   \n",
       "22  chr1        exon  15796  15947      -   \n",
       "23  chr1        exon  15005  15038      -   \n",
       "24  chr1        exon  14404  14501      -   \n",
       "25  chr1  transcript  14404  14501      -   \n",
       "26  chr1        exon  14404  14501      -   \n",
       "27  chr1  transcript  14404  24891      -   \n",
       "28  chr1        exon  24738  24891      -   \n",
       "29  chr1        exon  18268  18379      -   \n",
       "\n",
       "                                           attributes  \n",
       "12  gene_id \"ENSG00000227232.5\"; gene_name \"WASH7P...  \n",
       "13  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "14  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "15  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "16  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "17  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "18  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "19  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "20  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "21  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "22  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "23  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "24  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "25  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "26  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "27  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "28  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "29  gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# row = gtf.iloc[1]\n",
    "# import pprint\n",
    "# pprint.pprint(row.to_dict(), width=120)\n",
    "\n",
    "gtf.iloc[12:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "193d8b24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129911\n"
     ]
    }
   ],
   "source": [
    "# read the Parquet file\n",
    "df = pd.read_parquet(\"data/table/ENCFF180NXA.parquet\")\n",
    "\n",
    "# number of rows\n",
    "print(len(df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cb37824f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'chrom': 'chr16',\n",
      " 'end': 57591681,\n",
      " 'gene_id': 'ENSG00000159618.15',\n",
      " 'start': 57542421,\n",
      " 'strand': '+',\n",
      " 'transcripts': array([{'transcript_id': 'ENST00000615867.4', 'tx_start': 57542687, 'tx_end': 57565313, 'exon_starts': array([57542687, 57562056, 57562384, 57563091, 57563848, 57565034]), 'exon_ends': array([57542701, 57562157, 57562459, 57563247, 57563979, 57565313]), 'rel_abundance': 1.0}],\n",
      "      dtype=object)}\n",
      "Transcripts: 351270\n"
     ]
    }
   ],
   "source": [
    "row = df.iloc[18197]\n",
    "pprint.pprint(row.to_dict(), width=120)\n",
    "\n",
    "num_transcripts = df[\"transcripts\"].apply(len).sum()\n",
    "\n",
    "print(\"Transcripts:\", num_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f4efcb4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matching genes: 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id, chrom, start, end]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target = \"ENCLB376UFGT000206877\"\n",
    "\n",
    "matches = df[df[\"transcripts\"].apply(\n",
    "    lambda txs: any(t[\"transcript_id\"] == target for t in txs)\n",
    ")]\n",
    "\n",
    "print(\"Number of matching genes:\", len(matches))\n",
    "matches[[\"gene_id\", \"chrom\", \"start\", \"end\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23a976b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== SHAPES ===\n",
      "X_context       (10, 7, 256) torch.float16\n",
      "context_pos     (10, 7) torch.int64\n",
      "context_roles   (10, 7) torch.int64\n",
      "context_mask    (10, 7) torch.float32\n",
      "X_suffix        (10, 24918, 256) torch.float16\n",
      "suffix_pos      (10, 24918) torch.int64\n",
      "suffix_mask     (10, 24918) torch.float32\n",
      "target_idx      (10,) torch.int64\n",
      "target_role     (10,) torch.int64\n",
      "rel_abundance   (10,) torch.float32\n",
      "\n",
      "=== FIRST 3 CONTEXT POSITIONS ===\n",
      "tensor([[    0,     0,     0,     0,     0,     0,     0],\n",
      "        [19820,     0,     0,     0,     0,     0,     0],\n",
      "        [    0,     0,     0,     0,     0,     0,     0]])\n",
      "\n",
      "=== FIRST 3 CONTEXT ROLES ===\n",
      "tensor([[0, 0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0]])\n",
      "\n",
      "=== FIRST 3 TARGET IDX ===\n",
      "tensor([19820,    96,  9430])\n",
      "\n",
      "=== FIRST 3 TARGET ROLES ===\n",
      "tensor([0, 3, 0])\n",
      "\n",
      "=== FIRST 3 SUFFIX MASK SUMS ===\n",
      "tensor([24918.,  5097., 24918.])\n",
      "\n",
      "=== VERIFY SUFFIX START ===\n",
      "Sample 0: No context. suffix_pos starts at 0\n",
      "Sample 1: last ctx = 19820, suffix starts at 19821, diff = 1\n",
      "Sample 2: No context. suffix_pos starts at 0\n",
      "Sample 3: last ctx = 9430, suffix starts at 9431, diff = 1\n",
      "Sample 4: last ctx = 9583, suffix starts at 9584, diff = 1\n"
     ]
    }
   ],
   "source": [
    "# choose your parquet directory\n",
    "parquet_dir = \"data/table\"\n",
    "\n",
    "# get split filters\n",
    "train_filter, val_filter, test_filter = split_filters_by_chroms()\n",
    "\n",
    "# create dataset\n",
    "dset = GeneIterableFixedB(\n",
    "    parquet_dir=parquet_dir,\n",
    "    split_filter=train_filter,\n",
    "    tx_batch_size=10,\n",
    "    pad_bp=5000,\n",
    "    shuffle=False,   # deterministic for debugging\n",
    ")\n",
    "\n",
    "batch = next(iter(dset))\n",
    "\n",
    "# --------------------------------------------------------\n",
    "# 1. Print shapes\n",
    "# --------------------------------------------------------\n",
    "print(\"\\n=== SHAPES ===\")\n",
    "for k, v in batch.items():\n",
    "    if torch.is_tensor(v):\n",
    "        print(f\"{k:15s} {tuple(v.shape)} {v.dtype}\")\n",
    "    else:\n",
    "        print(f\"{k:15s} {type(v)}\")\n",
    "\n",
    "# --------------------------------------------------------\n",
    "# 2. Check the first 3 examples\n",
    "# --------------------------------------------------------\n",
    "print(\"\\n=== FIRST 3 CONTEXT POSITIONS ===\")\n",
    "print(batch[\"context_pos\"][:3])\n",
    "\n",
    "print(\"\\n=== FIRST 3 CONTEXT ROLES ===\")\n",
    "print(batch[\"context_roles\"][:3])\n",
    "\n",
    "print(\"\\n=== FIRST 3 TARGET IDX ===\")\n",
    "print(batch[\"target_idx\"][:3])\n",
    "\n",
    "print(\"\\n=== FIRST 3 TARGET ROLES ===\")\n",
    "print(batch[\"target_role\"][:3])\n",
    "\n",
    "print(\"\\n=== FIRST 3 SUFFIX MASK SUMS ===\")\n",
    "print(batch[\"suffix_mask\"][:3].sum(dim=1))\n",
    "\n",
    "# --------------------------------------------------------\n",
    "# 3. Check that the suffix starts at c+1\n",
    "# --------------------------------------------------------\n",
    "print(\"\\n=== VERIFY SUFFIX START ===\")\n",
    "ctx_pos = batch[\"context_pos\"]\n",
    "suffix_pos = batch[\"suffix_pos\"]\n",
    "for i in range(5):\n",
    "    ctx_len = (batch[\"context_mask\"][i] > 0).sum().item()\n",
    "    if ctx_len == 0:\n",
    "        print(f\"Sample {i}: No context. suffix_pos starts at {suffix_pos[i,0].item()}\")\n",
    "    else:\n",
    "        last_ctx = ctx_pos[i, ctx_len-1].item()\n",
    "        suf0 = suffix_pos[i, 0].item()\n",
    "        print(f\"Sample {i}: last ctx = {last_ctx}, suffix starts at {suf0}, diff = {suf0 - last_ctx}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5e7f65cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import random\n",
    "from pathlib import Path\n",
    "\n",
    "import pyarrow as pa\n",
    "import pyarrow.dataset as ds\n",
    "import torch\n",
    "\n",
    "ID2ROLE = [\"TSS\", \"D\", \"A\", \"PAS\"]\n",
    "\n",
    "def validate_batch_basic(batch, verbose=True):\n",
    "    X_context     = batch[\"X_context\"]\n",
    "    context_pos   = batch[\"context_pos\"]\n",
    "    context_roles = batch[\"context_roles\"]\n",
    "    context_mask  = batch[\"context_mask\"]\n",
    "    X_suffix      = batch[\"X_suffix\"]\n",
    "    suffix_pos    = batch[\"suffix_pos\"]\n",
    "    suffix_mask   = batch[\"suffix_mask\"]\n",
    "    target_idx    = batch[\"target_idx\"]\n",
    "    target_role   = batch[\"target_role\"]\n",
    "    rel_ab        = batch[\"rel_abundance\"]\n",
    "\n",
    "    B, M, D = X_context.shape\n",
    "    B2, S, D2 = X_suffix.shape\n",
    "\n",
    "    assert B == B2, \"Batch size mismatch between context and suffix\"\n",
    "    assert D == D2, \"Embedding dims mismatch between context and suffix\"\n",
    "    assert context_pos.shape == (B, M,)\n",
    "    assert context_roles.shape == (B, M,)\n",
    "    assert context_mask.shape == (B, M,)\n",
    "    assert suffix_pos.shape == (B, S,)\n",
    "    assert suffix_mask.shape == (B, S,)\n",
    "    assert target_idx.shape == (B,)\n",
    "    assert target_role.shape == (B,)\n",
    "    assert rel_ab.shape == (B,)\n",
    "\n",
    "    # masks must be prefix-ones then zeros\n",
    "    ctx_lengths = (context_mask > 0).sum(dim=1)\n",
    "    suf_lengths = (suffix_mask > 0).sum(dim=1)\n",
    "\n",
    "    for i in range(B):\n",
    "        ctx_len = ctx_lengths[i].item()\n",
    "        suf_len = suf_lengths[i].item()\n",
    "        if ctx_len > 0:\n",
    "            assert torch.all(context_mask[i, :ctx_len] == 1)\n",
    "            assert torch.all(context_mask[i, ctx_len:] == 0)\n",
    "        if suf_len > 0:\n",
    "            assert torch.all(suffix_mask[i, :suf_len] == 1)\n",
    "            assert torch.all(suffix_mask[i, suf_len:] == 0)\n",
    "\n",
    "    if verbose:\n",
    "        print(\"✓ basic shapes and masks OK\")\n",
    "        print(f\"  B={B}, M={M}, S={S}, D={D}\")\n",
    "        print(\"  mean ctx length:\", ctx_lengths.float().mean().item())\n",
    "        print(\"  mean suffix length:\", suf_lengths.float().mean().item())\n",
    "\n",
    "\n",
    "def validate_suffix_alignment(batch, verbose=True):\n",
    "    context_pos  = batch[\"context_pos\"]\n",
    "    context_mask = batch[\"context_mask\"]\n",
    "    suffix_pos   = batch[\"suffix_pos\"]\n",
    "    suffix_mask  = batch[\"suffix_mask\"]\n",
    "\n",
    "    B, M = context_pos.shape\n",
    "    _, S = suffix_pos.shape\n",
    "\n",
    "    ctx_lengths = (context_mask > 0).sum(dim=1)\n",
    "    suf_lengths = (suffix_mask > 0).sum(dim=1)\n",
    "\n",
    "    for i in range(B):\n",
    "        ctx_len = ctx_lengths[i].item()\n",
    "        suf_len = suf_lengths[i].item()\n",
    "        if suf_len == 0:\n",
    "            continue\n",
    "\n",
    "        if ctx_len == 0:\n",
    "            # No context: suffix should start at 0\n",
    "            assert suffix_pos[i, 0].item() == 0, f\"Sample {i}: no context but suffix_pos[0] != 0\"\n",
    "        else:\n",
    "            last_ctx = context_pos[i, ctx_len - 1].item()\n",
    "            suf0 = suffix_pos[i, 0].item()\n",
    "            assert suf0 == last_ctx + 1, f\"Sample {i}: suffix start {suf0} != last_ctx+1 {last_ctx+1}\"\n",
    "\n",
    "        # Check monotonicity inside suffix\n",
    "        sufp = suffix_pos[i, :suf_len]\n",
    "        assert torch.all(sufp[1:] > sufp[:-1]), f\"Sample {i}: suffix positions not strictly increasing\"\n",
    "\n",
    "    if verbose:\n",
    "        print(\"✓ suffix alignment and monotonicity OK\")\n",
    "\n",
    "\n",
    "def validate_target_alignment(batch, verbose=True):\n",
    "    suffix_mask = batch[\"suffix_mask\"]\n",
    "    suffix_pos  = batch[\"suffix_pos\"]\n",
    "    target_idx  = batch[\"target_idx\"]\n",
    "\n",
    "    B, S = suffix_mask.shape\n",
    "    suf_lengths = (suffix_mask > 0).sum(dim=1)\n",
    "\n",
    "    for i in range(B):\n",
    "        suf_len = suf_lengths[i].item()\n",
    "        ti = target_idx[i].item()\n",
    "        assert 0 <= ti < suf_len, f\"Sample {i}: target_idx {ti} out of suffix length {suf_len}\"\n",
    "\n",
    "    if verbose:\n",
    "        print(\"✓ target_idx in-range for all samples\")\n",
    "\n",
    "\n",
    "def pretty_print_small_batch(batch, n=5):\n",
    "    n = min(n, batch[\"X_context\"].shape[0])\n",
    "    ctx_pos   = batch[\"context_pos\"]\n",
    "    ctx_mask  = batch[\"context_mask\"]\n",
    "    ctx_roles = batch[\"context_roles\"]\n",
    "    suf_pos   = batch[\"suffix_pos\"]\n",
    "    suf_mask  = batch[\"suffix_mask\"]\n",
    "    tgt_idx   = batch[\"target_idx\"]\n",
    "    tgt_role  = batch[\"target_role\"]\n",
    "    rel_ab    = batch[\"rel_abundance\"]\n",
    "\n",
    "    print(\"\\n=== SMALL BATCH SUMMARY ===\")\n",
    "    for i in range(n):\n",
    "        ctx_len = int(ctx_mask[i].sum().item())\n",
    "        suf_len = int(suf_mask[i].sum().item())\n",
    "        print(f\"\\nSample {i}\")\n",
    "        print(\"  rel_abundance:\", float(rel_ab[i].item()))\n",
    "        print(\"  ctx_len:\", ctx_len, \"suffix_len:\", suf_len)\n",
    "        print(\"  context_pos:\", ctx_pos[i, :ctx_len].tolist())\n",
    "        print(\"  context_roles:\", [ID2ROLE[int(x)] for x in ctx_roles[i, :ctx_len]])\n",
    "        print(\"  suffix_pos[0:5]:\", suf_pos[i, :min(5, suf_len)].tolist())\n",
    "        print(\"  target_idx:\", int(tgt_idx[i].item()))\n",
    "        print(\"  target_role:\", ID2ROLE[int(tgt_role[i].item())])\n",
    "\n",
    "\n",
    "def validate_one_batch(batch, verbose=True):\n",
    "    validate_batch_basic(batch, verbose=verbose)\n",
    "    validate_suffix_alignment(batch, verbose=verbose)\n",
    "    validate_target_alignment(batch, verbose=verbose)\n",
    "    if verbose:\n",
    "        pretty_print_small_batch(batch, n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f58a81ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ basic shapes and masks OK\n",
      "  B=16, M=5, S=12525, D=256\n",
      "  mean ctx length: 2.25\n",
      "  mean suffix length: 8011.25\n",
      "✓ suffix alignment and monotonicity OK\n",
      "✓ target_idx in-range for all samples\n",
      "\n",
      "=== SMALL BATCH SUMMARY ===\n",
      "\n",
      "Sample 0\n",
      "  rel_abundance: 0.1875\n",
      "  ctx_len: 0 suffix_len: 12525\n",
      "  context_pos: []\n",
      "  context_roles: []\n",
      "  suffix_pos[0:5]: [0, 1, 2, 3, 4]\n",
      "  target_idx: 5000\n",
      "  target_role: TSS\n",
      "\n",
      "Sample 1\n",
      "  rel_abundance: 0.1875\n",
      "  ctx_len: 1 suffix_len: 7524\n",
      "  context_pos: [5000]\n",
      "  context_roles: ['TSS']\n",
      "  suffix_pos[0:5]: [5001, 5002, 5003, 5004, 5005]\n",
      "  target_idx: 166\n",
      "  target_role: D\n",
      "\n",
      "Sample 2\n",
      "  rel_abundance: 0.1875\n",
      "  ctx_len: 2 suffix_len: 7357\n",
      "  context_pos: [5000, 5167]\n",
      "  context_roles: ['TSS', 'D']\n",
      "  suffix_pos[0:5]: [5168, 5169, 5170, 5171, 5172]\n",
      "  target_idx: 468\n",
      "  target_role: A\n",
      "\n",
      "Sample 3\n",
      "  rel_abundance: 0.1875\n",
      "  ctx_len: 3 suffix_len: 6888\n",
      "  context_pos: [5000, 5167, 5636]\n",
      "  context_roles: ['TSS', 'D', 'A']\n",
      "  suffix_pos[0:5]: [5637, 5638, 5639, 5640, 5641]\n",
      "  target_idx: 144\n",
      "  target_role: D\n",
      "\n",
      "Sample 4\n",
      "  rel_abundance: 0.1875\n",
      "  ctx_len: 4 suffix_len: 6743\n",
      "  context_pos: [5000, 5167, 5636, 5781]\n",
      "  context_roles: ['TSS', 'D', 'A', 'D']\n",
      "  suffix_pos[0:5]: [5782, 5783, 5784, 5785, 5786]\n",
      "  target_idx: 301\n",
      "  target_role: A\n"
     ]
    }
   ],
   "source": [
    "parquet_dir = \"data/table\"\n",
    "train_filter, val_filter, test_filter = split_filters_by_chroms()\n",
    "\n",
    "dset = GeneIterableFixedB(\n",
    "    parquet_dir=parquet_dir,\n",
    "    split_filter=train_filter,\n",
    "    tx_batch_size=16,\n",
    "    pad_bp=5000,\n",
    "    shuffle=True,\n",
    ")\n",
    "\n",
    "batch = next(iter(dset))\n",
    "validate_one_batch(batch, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe9de2b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "isoform-gen",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
